### PR TITLE
Filesystem

### PR DESCRIPTION
**Pull request by HerrCooker**
Merge FileSystem into develop branch

### CODE REVIEW

File: src/filesystem/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,9 @@
+add_library(file_system INTERFACE
```
Comment (DavidFederl) [inline src/filesystem/CMakeLists.txt:1]: Please rename the library to `filesystem_lib`

File: src/filesystem/FileSystem.c [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+
+#include "Common.h"
+#include "EnV5HwConfiguration.h"
+#include "FileSystem.h"
+#include "FileSystemInternal.h"
+#include "Flash.h"
+#include "middleware.h"
+
+#include <stdlib.h>
+#include <string.h>
```
Comment (DavidFederl) [inline src/filesystem/FileSystem.c:11]:
```suggestion
#define SOURCE_FILE "FILESYSTEM"

#include <math.h>
#include <stdlib.h>
#include <string.h>

#include "Common.h"
#include "EnV5HwConfiguration.h"
#include "FileSystem.h"
#include "FileSystemInternal.h"
#include "Flash.h"
#include "middleware.h"
```

File: src/filesystem/FileSystem.c [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+uint8_t numberOfEntries = 0;
+uint8_t sectorFree[10];
+uint8_t currentID = 100;
+
+uint32_t nextFileSystemSector;
+fileSystemEntry fileSystem[60];
+size_t fileSystemLength = sizeof(fileSystem) / sizeof(fileSystemEntry);
+uint8_t entrySize = 8;
+
+uint32_t nextFileSector = 0;
```
Comment (DavidFederl) [inline src/filesystem/FileSystem.c:22]: make variables for internal usage `static`

File: test/hardware/TestFlashFileSystem/DemoFileSystem.c [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+
+#include "Common.h"
+#include "EnV5HwConfiguration.h"
+#include "EnV5HwController.h"
+#include "Esp.h"
+#include "FileSystem.h"
+#include "Flash.h"
+#include "FpgaConfigurationHandler.h"
+#include "Network.h"
+
```
Comment (DavidFederl) [inline test/hardware/TestFlashFileSystem/DemoFileSystem.c:35]:
```suggestion
```

File: test/hardware/TestFlashFileSystem/DemoFileSystem.c [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+- always use findFittingStartSector() to check if your file fits. If it does, the sector to be used
+is stored in the variable nextFileSector
+- nextFileSector is the number of the sector you should write your file to. This is to be converted
+  into an actual address.
+
+*/
+
+#define SOURCE_FILE "CONFIGURE-HWTEST"
+
+#include "stdlib.h"
```
Comment (DavidFederl) [inline test/hardware/TestFlashFileSystem/DemoFileSystem.c:16]:
```suggestion
#include <stdlib.h>
```

File: test/hardware/TestFlashFileSystem/DemoFileSystem.c
Comment (DavidFederl) [inline test/hardware/TestFlashFileSystem/DemoFileSystem.c:1]: Please replace `printf(...)` calls with calls to `PRINT(...)` from `Common.h`!

File: src/filesystem/include/FileSystem.h [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    uint8_t numberOfSectors;
+} fileSystemEntry;
+
+extern fileSystemEntry fileSystem[];
+extern size_t fileSystemLength;
+extern uint8_t numberOfEntries;
+extern uint8_t sectorFree[10];
+extern uint8_t currentID;
+
+extern uint32_t nextFileSector;
```
Comment (DavidFederl) [inline src/filesystem/include/FileSystem.h:23]: Are you sure access to these variables is necessary from outside the library?

File: src/filesystem/include/FileSystem.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
+#ifndef FILESYSTEM_H
+#define FILESYSTEM_H
+
+#include <FlashTypedefs.h>
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
```
Comment (DavidFederl) [inline src/filesystem/include/FileSystem.h:7]:
```suggestion
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>

#include "FlashTypedefs.h"
```

File: src/filesystem/include/FileSystem.h [hunk 8]
```diff
@@ -0,0 +1,39 @@
+#ifndef FILESYSTEM_H
+#define FILESYSTEM_H
```
Comment (DavidFederl) [inline src/filesystem/include/FileSystem.h:2]:
```suggestion
#define ENV5_FILESYSTEM_HEADER
```

File: src/filesystem/include/FileSystem.h [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+void moveFileToSector(flashConfiguration_t *flashConfig, uint8_t ID, uint8_t newSector);
+bool eraseFileByID(flashConfiguration_t *flashConfig, uint8_t id);
+void printFileSystem();
+
+uint8_t getNumberOfFreeSectors();
+fileSystemEntry *getEntryByID(uint8_t id);
+fileSystemEntry *getEntryByIndex(uint8_t index);
+fileSystemEntry *getEntryBySector(uint8_t sector);
+void sortFileSystemByStartSector();
+void sortFileSystemByID();
```
Comment (DavidFederl) [inline src/filesystem/include/FileSystem.h:37]: Please prepend all functions of public API with `filesystem`, e.g. `filesystemInit` / `filesystemFindFittingStartSector` / ...

File: src/filesystem/FileSystemInternal.h [hunk 10]
```diff
@@ -0,0 +1,22 @@
+#ifndef FILESYSTEMINTERNAL_H
+#define FILESYSTEMINTERNAL_H
```
Comment (DavidFederl) [inline src/filesystem/FileSystemInternal.h:2]:
```suggestion
#define ENV5_FILESYSTEM_INTERNAL_HEADER
```

File: src/filesystem/FileSystem.c
Comment (DavidFederl) [inline src/filesystem/FileSystem.c:1]: Please replace all `printf(...)` with `PRINT_DEBUG(...)` calls from `Common.h`

File: src/filesystem/FileSystem.c [hunk 11]
```diff
@@ -0,0 +1,502 @@
...
+}
+
+void addNewFileSystemEntry(flashConfiguration_t *flashConfig, uint32_t size, uint8_t isConfig) {
+    uint8_t sector = nextFileSector;
+    uint8_t id = currentID;
+    currentID++;
+    fileSystem[numberOfEntries].id = id;
+    /*for (int i = 0; i <= sizeof(*name); i++) {
+        //fileSystem[numberOfEntries].name[i] = name[i];
+    }*/
```
Comment (DavidFederl) [inline src/filesystem/FileSystem.c:89]: Remove if not used!

**Review by DavidFederl — CHANGES_REQUESTED** [review]
Feel free to ask if you have problems.

File: test/hardware/TestFlashFileSystem/DemoFileSystem.c (hunk 4 above)
Comment (HerrCooker) [reply test/hardware/TestFlashFileSystem/DemoFileSystem.c:35]: I'm not sure why these should be removed, as they are needed.

Comment (DavidFederl) [reply test/hardware/TestFlashFileSystem/DemoFileSystem.c:35]: Only the not required empty line between the `#include` calls

File: src/filesystem/include/FileSystem.h [hunk 12]
```diff
@@ -7,16 +7,16 @@
 #include "FlashTypedefs.h"
 
 typedef union fileSystemEntry {
-    //check new union for constintency
+    // check new union for constintency
```
Comment (DavidFederl) [inline src/filesystem/include/FileSystem.h:10]:
```suggestion
```

**Review by DavidFederl — CHANGES_REQUESTED** [review]

**Review by DavidFederl — APPROVED** [review]

**Pull request merged**